src/bigint.rs

```diff
@@ -467,146 +522,303 @@ impl Unsigned for BigUint {}
 
 forward_all_binop_to_val_ref_commutative!(impl Add for BigUint, add);
 
-impl<'a> Add<&'a BigUint> for BigUint {
+fn add2(a: &mut [BigDigit], b: &[BigDigit]) {
+    let mut b_iter = b.iter();
+    let mut carry = 0;
+
+    for ai in a.iter_mut() {
+        if let Some(bi) = b_iter.next() {
+            *ai = adc(*ai, *bi, &mut carry);
+        } else if carry != 0 {
+            *ai = adc(*ai, 0, &mut carry);
+        } else {
+            break;
+        }
+    }
+
+    assert!(b_iter.next() == None && carry == 0);
+}
+
+impl<'a> Add<&'a [BigDigit]> for BigUint {
```

Comment:
This new impl is visible in the public API, which is undesirable.  AFAICS this only used by the normal Add implementation, so let's keep it as one.